### PR TITLE
GODRIVER-2722 Remove/unexport unnecessary interfaces.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-2722

## Summary
Remove/un-export unnecessary interfaces.

### CODE REVIEW

**Review by prestonvasquez — APPROVED** [review]

File: bson/primitive.go [hunk 1]
```diff
@@ -16,10 +16,10 @@ import (
...
 )
 
-// Zeroer allows custom struct types to implement a report of zero
-// state. All struct types that don't implement Zeroer or where IsZero
+// zeroer allows custom struct types to implement a report of zero
+// state. All struct types that don't implement zeroer or where IsZero
 // returns false are considered to be not zero.
-type Zeroer interface {
+type zeroer interface {
 	IsZero() bool
```
Comment (matthewdale) [inline bson/primitive.go:23]: Considering the recent relevance of `IsZero`, it's possible we may need to keep the `Zeroer` interface exported.

File: bson/proxy.go [hunk 2]
```diff
@@ -6,9 +6,9 @@
 
 package bson
 
-// Proxy is an interface implemented by types that cannot themselves be directly encoded. Types
+// proxy is an interface implemented by types that cannot themselves be directly encoded. Types
 // that implement this interface with have ProxyBSON called during the encoding process and that
 // value will be encoded in place for the implementer.
-type Proxy interface {
+type proxy interface {
 	ProxyBSON() (interface{}, error)
```
Comment (matthewdale) [inline bson/proxy.go:13]: We should consider completely removing the `ProxyBSON` feature. Based on [searching for uses in open-source repos](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+-file:%28%5E%7C/%29vendor/+ProxyBSON&patternType=standard&case=yes&sm=0&groupBy=repo), it seems completely unused.

File: bson/struct_codec.go [hunk 3]
```diff
@@ -181,7 +181,9 @@ func (sc *StructCodec) EncodeValue(ec EncodeContext, vw ValueWriter, val reflect
 		encoder := desc.encoder
 
 		var empty bool
-		if cz, ok := encoder.(CodecZeroer); ok {
+		if cz, ok := encoder.(interface {
+			IsTypeZero(interface{}) bool
+		}); ok {
 			empty = cz.IsTypeZero(rv.Interface())
```
Comment (matthewdale) [inline bson/struct_codec.go:187]: We should consider completely removing the `IsTypeZero` feature. Based on [searching for uses in open-source repos](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+-file:%28%5E%7C/%29vendor/+IsTypeZero&patternType=standard&case=yes&sm=0&groupBy=repo), it seems completely unused.

**Review by prestonvasquez — APPROVED** [review]

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

**Review by matthewdale — APPROVED** [review]

**Pull request merged**